rust/src/adapter/mod.rs

```diff
@@ -1,122 +1,101 @@
-pub mod otel_formatter;
+use tokio::sync::mpsc::{Sender, channel};
+use anyhow::Result;
+use log::warn;
+use wasmtime::Linker;
+
+use crate::{
+    Event,
+    TelemetryId,
+    context::add_to_linker,
+    collector::{Collector, CollectorHandle}, TraceEvent
+};
+
 pub mod otelstdout;
-pub mod stdout;
+pub mod otel_formatter;
 pub mod datadog;
 pub mod datadog_formatter;
-pub mod zipkin_formatter;
 pub mod zipkin;
+pub mod zipkin_formatter;
+pub mod stdout;
 
-use core::time;
-use std::{sync::Arc, thread};
-use std::sync::atomic::{AtomicUsize, Ordering};
-
-use rand::Rng;
-use anyhow::Result;
-use tokio::sync::{mpsc::Sender, Mutex};
-
-use crate::{Event, EventChannel, Metadata};
-
-#[derive(Debug, Clone)]
-pub struct TelemetryId(u128);
-
-impl TelemetryId {
-    /// format as 8-byte zero-prefixed hex string
-    pub fn to_hex_8(&self) -> String {
-        // 8 bytes is 16 chars
-        format!("{:016x}", self.0 as u64)
-    }
-    /// format as 16-byte zero-prefixed hex string
-    pub fn to_hex_16(&self) -> String {
-        // 16 bytes is 32 chars
-        format!("{:032x}", self.0)
-    }
-}
 
-impl From<TelemetryId> for u64 {
-    fn from(v: TelemetryId) -> Self {
-        v.0 as u64
-    }
-}
+/// An adapter represents a sink for events and is mostly implementation specific
+/// to the sink that the data is being sent to and the format that the data is in.
+/// Collectors batch up events and metadata into a TraceEvent and then sends this
+/// payload to the Adapter when the module is done running. When implementing an adapter,
+/// your job is to handle this event by implementing the function handle_trace_event.
+///
+/// ┌────────────────┐               ┌──────────────┐                ┌────────────────┐
+/// │   Collector    │               │    Adapter   │                │   Collector    │
+/// │                │               │              │                │                │
+/// │                │               │              │                │                │
+/// ├────────────────┤  <TraceEvent> ├──────────────┤  <TraceEvent>  ├────────────────┤
+/// │  adapter_tx    ├──────────────►│ adapter_rx   │◄───────────────┤  adapter_tx    │
+/// └────────────────┘               └──────────────┘                └────────────────┘
+pub trait Adapter {
+    /// Callback which is used when a TraceEvent is sent from the Collector
+    /// This must be implemented by an Adapter
+    fn handle_trace_event(&mut self, evt: TraceEvent) -> Result<()>;
```

Comment:
Yeah good point. I think it could be an enum or something too. Let me think about it though. I think some adapters, like datadog, will want the metrics and logs bounded by the trace, and not treated separately.